analysis/src/SharedTypes.ml

```diff
@@ -296,6 +294,36 @@ end
 
 type polyVariantConstructor = {name: string; args: Types.type_expr list}
 
+(** An type that can be used to drive completion *)
+type completionType =
+  | Tuple of QueryEnv.t * Types.type_expr list * Types.type_expr
+  | Toption of QueryEnv.t * completionType
+  | Tbool of QueryEnv.t
+  | Tarray of QueryEnv.t * completionType
```

Comment:
Looking at `env` in the nested case, e.g. in the argument to `Tarray`.
It seems to me, from a cursory look, that the inner env might always be the same as the outer env.
If true in general, then factoring out the `env` in the type definition means it does not need to appear in the argument of `Tarray` etc.

I haven't really checked that this is true for all cases. If it is true, then the overall type would be:
`{env; desc}` where `desc` would be the recursive definition of the variant cases. So e.g. `Tarray of desc` would not include the `env`.